{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3a272ce4-c003-4d6a-9ecc-16279d05d048",
   "metadata": {},
   "source": [
    "# Generation of Hi-C map by Juicer\n",
    "\n",
    "> Modified on Nov 24th, 2021 by Meng Zhang (mzhang@mpipz.mpg.de)\n",
    "\n",
    "Juicer is a powerful tool for Hi-C data processing and analysis. Here we just take the advantage of the pipeline that creates Hi-C files from raw data. If you are interested in the complete function of juicer, please visit the GitHub of [Aidenlab](https://github.com/aidenlab/juicer).\n",
    "\n",
    "## Step 1: Install Juicer\n",
    "You can also go to [juicer installation page](https://github.com/aidenlab/juicer/wiki/Installation#quick-start) to read detailed guidance and interpretation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bde36bb-8a8d-4fa5-9f94-a8ab126b54c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /path/to/install/juicer/\n",
    "git clone https://github.com/theaidenlab/juicer.git\n",
    "\n",
    "# create a symbolic link in working directory\n",
    "cd /path/to/your/HiC/analysis/working/directory/\n",
    "ln -s ~/juicer/CPU scripts\n",
    "\n",
    "# juicer_tools is required in Hi-C map generation step\n",
    "# take CPU version as example\n",
    "cd scripts/common\n",
    "wget https://hicfiles.tc4ga.com/public/juicer/juicer_tools.1.9.9_jcuda.0.8.jar\n",
    "ln -s juicer_tools.1.9.9_jcuda.0.8.jar  juicer_tools.jar"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55073db3-f986-48eb-95bf-5855b4efb411",
   "metadata": {},
   "source": [
    "## Step 2: Configrue directory structure\n",
    "\n",
    "Juicer requires a particular directory structure to navigate. A typical sample is given below, and it is highly recommended you configure your files and directories as the same structure to ensure that juicer is able to find files properly. [Juicer wiki](https://github.com/aidenlab/juicer/wiki/Installation#quick-start) also gives a clear interpretation about the directory structure.\n",
    "\n",
    "![Recommended directory structure](./juicer_directory_structure.png)\n",
    "\n",
    "All files associated with juicer analysis are suggested to be put under one directory. `scripts` is a symbolic link to where the main juicer scripts locate, which might be different for CPU and cluster versions. Please determine your running platform before creating the link, and also refer to the examples given by [juicer wiki](https://github.com/aidenlab/juicer/wiki/Installation#quick-start). Note that the above installation command takes CPU version as example. `reference/` folder stores the reference genome and its BWA indexed files. `restriction_sites/` contains the restriction site files, see below to learn about how to generate them. The sequenced Hi-C reads are in `fastq/`, which can remain gzipped. Files generated by juicer will be stored in `aligned/` and `split/` but you do not have to create these two folders because they will be created when you run juicer."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84834c71-de04-425a-a61f-597c5b61483e",
   "metadata": {},
   "source": [
    "## Step 3: Preprocessing\n",
    "If reference genome has not been indexed, please use BWA to index under `reference/` directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9454dc71-60e9-4b27-b8db-ce87d2f099eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "JSD='/path/to/juicer/scripts/directory' # directory where you put scripts, i.e. your juicer analysis working directory\n",
    "REF='/path/to/reference/reference.fasta' \n",
    "genomeID='your_reference_name' # should be the same as reference genome name, that is \n",
    "thread=<number of threads>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57b191fb-2685-495b-9810-628972acdd2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# index reference genome\n",
    "cd $JSD/reference\n",
    "bwa index $REF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99d611b9-1a90-46aa-9bf1-f119909bb09e",
   "metadata": {},
   "source": [
    "Python program `generate_site_positions.py` is more stable when running via python2, so if unexpected errors appear, please inspect your running environment. But this problem could possibly be solved in latest juicer version. Plus, note the restriction enzyme used when generating Hi-C data. If it is not defined in `generate_site_positions.py`, click [here](https://github.com/aidenlab/juicer/wiki/Usage) to learn about how to modify the script or set juicer flag. If you don't want to do any fragment-level analysis (as with a DNAse experiment), you should assign the site \"none\", as in `juicer.sh -s none`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "557d660a-dec4-4f2b-a789-a7a93bb4fc17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create restriction sites file\n",
    "cd $JSD/restriction_sites\n",
    "python /path/to/juicer/misc/generate_site_positions.py \\\n",
    "DpnII  $genomeID $REF\n",
    "\n",
    "# generate genome.chrom.sizes\n",
    "awk 'BEGIN{OFS=\"\\t\"}{print $1, $NF}' ${genomeID}_DpnII.txt \\\n",
    " > ${genomeID}.chrom.sizes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b6384b8-db93-4809-9e14-351443d06cd5",
   "metadata": {},
   "source": [
    "## Step 4: Run juicer for each replicate\n",
    "\n",
    "If Hi-C sequencing data consists of multiple replicates, no matter biological or technical, it is suggested to regard them as independent works, and run juicer separately. For example, the Hi-C data I used came from two different libraries (1M and 2M, biological replicates), and two different PCRs amplification based on each library (R1 and R2, technical replicates), so it ended up that four alignment works ran separately."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0648a820-4515-49c3-a1d7-6838b81a5dc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# juicer pipeline for each replicate\n",
    "cd $JSD\n",
    "$JSD/scripts/juicer.sh -g $genomeID -s DpnII -t 64 \\\n",
    " -d $JSD/HiCwork/replicate1/  -D $JSD  -z $REF \\\n",
    " -p $JSD/restriction_sites/${genomeID}.chrom.sizes \\\n",
    " -y $JSD/restriction_sites/${genomeID}_DpnII.txt\n",
    " \n",
    "$JSD/scripts/juicer.sh -g $genomeID -s DpnII -t 64 \\\n",
    " -d $JSD/HiCwork/replicate2/  -D $JSD  -z $REF \\\n",
    " -p $JSD/restriction_sites/${genomeID}.chrom.sizes \\\n",
    " -y $JSD/restriction_sites/${genomeID}_DpnII.txt\n",
    "\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9660f7dc-2535-465d-aacf-59456324535e",
   "metadata": {},
   "source": [
    "## Step 5: Merge replicates (optional)\n",
    "\n",
    "If your Hi-C sequencing data are from multiple replicates, it is necessary to merge them to generate a combined Hi-C map. To this end, Juicer provides us `MEGA`. You can find detailed interpretaion in [juicer manual](https://github.com/aidenlab/juicer/wiki/Usage#creating-a-mega-map). However, if you do not have sequencing replicates, which means only one juicer analysis is implemented in step 4, this step is not required."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f203ef8-1e85-4621-a985-b8d02dd7b878",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd $JSD/HiC_work\n",
    "$JSD/scripts/common/mega.sh -D $JSD -g $gneomeID -s DpnII"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "776bb1c7-7306-43e2-9249-68d87d7ab050",
   "metadata": {},
   "source": [
    "In the end, you can find the 'mega' map and some statistics at `$JSD/HiC_work/mega/aligned/`. It is worth to note that juicer produces Hi-C map with quality over 1 and 30 They are named as `inter.hic` and `inter_30.hic`, and the corresponding statistics are named as `inter.txt` and `inter_30.txt`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "643f202d-9ac3-4dcd-be2e-35959078495e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
